eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,404 @@
+		if int(nothingToSyncCount.Load()) == len(ec.clients) {
+			// All clients returned ErrNothingToSync
+			return nil, nil, ErrNothingToSync
+		}
+		return nil, nil, err
+	}
+
+	go func() {
+		defer func() {
+			close(logsCh)
+			close(errCh)
+		}()
+
+		var needInitChans atomic.Bool
+
+		// Getting historical logs may take significant amount of time.
+		// The current client may fail in the middle of the process,
+		// and also some clients may become available again, therefore we need to try other clients on error.
+		// Since singleLogsCh and singleErrCh are initialized by the active client,
+		// we need to re-initialize them if we try other clients.
+		processLogsFunc := func(client SingleClientProvider) (any, error) {
+			if needInitChans.Load() {
+				if _, err := startFetchingFunc(client); err != nil {
+					return nil, err
+				}
+			} else {
+				needInitChans.Store(true)
+			}
+
+			var lastBlock uint64
+			for {
+				select {
+				case <-ctx.Done():
+					return nil, ctx.Err()
+				case <-ec.closed:
+					return nil, fmt.Errorf("client closed")
+				case log, ok := <-singleLogsCh:
+					if !ok {
+						// Underlying channel is closed -> no more logs.
+						return nil, nil
+					}
+					logsCh <- log
+					lastBlock = max(lastBlock, log.BlockNumber)
+				case err, ok := <-singleErrCh:
+					if !ok {
+						// If the error channel closed, treat that as no more logs or success.
+						return nil, nil
+					}
+					fromBlock = max(fromBlock, lastBlock+1)
+					return nil, err // Try another client
+				}
+			}
+		}
+
+		if _, err := ec.call(ec.setMethod(ctx, "FetchHistoricalLogs [process logs]"), processLogsFunc); err != nil {
+			errCh <- err
+		}
+	}()
+
+	return logsCh, errCh, nil
+}
+
+// StreamLogs subscribes to events emitted by the contract.
+func (ec *MultiClient) StreamLogs(ctx context.Context, fromBlock uint64) <-chan BlockLogs {
+	logs := make(chan BlockLogs)
+
+	go func() {
+		defer close(logs)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ec.closed:
+				return
+			default:
+				f := func(client SingleClientProvider) (any, error) {
+					lastBlock, err := client.streamLogsToChan(ctx, logs, fromBlock)
+					if errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled) {
+						// Closed gracefully.
+						return lastBlock, err
+					}
+					if err != nil {
+						fromBlock = max(fromBlock, lastBlock+1)
+						return nil, err
+					}
+
+					// Success: terminate the loop without error
+					return nil, nil
+				}
+
+				_, err := ec.call(ec.setMethod(ctx, "StreamLogs"), f)
+				if errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled) {
+					// Closed gracefully.
+					return
+				}
+				if err != nil {
+					ec.logger.Fatal("failed to stream registry events", zap.Error(err))
+					return // Terminate the loop after logging Fatal
+				}
+				// On success, terminate the loop
+				return
```

Comment:
honestly I don't know if this is simpler or clearer, just shorter.. let me know what you think
```suggestion
				if err !=nil && !(errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled)) {
					ec.logger.Fatal("failed to stream registry events", zap.Error(err))
				}
				return
```

Reply:
I don't think we should call `Fatal` on client closure or context cancellation because they should indicate "successful closure", e.g. graceful shutdown if the pod is restarted. There won't be probably any difference in actual behavior but I'd prefer to keep process exit codes clean

Reply:
Sorry, I didn't read your suggestion well enough, so my previous message is wrong. Yes, we can do what you suggest

Reply:
Changed